Expose a Register function that can be used to install routes in subrouter

maddy-jo opened the pull request:
Expose a Register function that can be used to install routes in subrouter

Takes the existing logic for generating the routes in gen-go/server and applies them to an injected router. This allows a server to compose routers by calling

```go
Register(parentRouter.PathPrefix("/subrouter-prefix").Subrouter(), controller)
```

Testing this hypothesis out in app-district-service.

## Jira

[SHAPI-1697](https://clever.atlassian.net/browse/SHAPI-1697)

## Testing

See Clever/app-district-service#2.

```
$ curl -X GET http://localhost:8080/v2/things -i          
HTTP/1.1 500 Internal Server Error
Content-Type: text/plain; charset=utf-8
Vary: Accept-Encoding
X-Content-Type-Options: nosniff
Date: Tue, 17 Jun 2025 00:22:25 GMT
Content-Length: 44

{"message":"TODO, this is just an example"}
madeline.lumetta> app-district-service $ curl -X GET http://localhost:8080/districts/v2/things -i
HTTP/1.1 500 Internal Server Error
Content-Type: text/plain; charset=utf-8
Vary: Accept-Encoding
X-Content-Type-Options: nosniff
Date: Tue, 17 Jun 2025 00:22:30 GMT
Content-Length: 44

{"message":"TODO, this is just an example"}
madeline.lumetta> app-district-service $ curl -X GET http://localhost:8080/sessions/v2/things -i 
HTTP/1.1 500 Internal Server Error
Content-Type: text/plain; charset=utf-8
Vary: Accept-Encoding
X-Content-Type-Options: nosniff
Date: Tue, 17 Jun 2025 00:22:38 GMT
Content-Length: 44

{"message":"TODO, this is just an example"}

madeline.lumetta> app-district-service $ curl -X GET http://localhost:8080/wrong/v2/things -i   
HTTP/1.1 404 Not Found
Content-Type: text/plain; charset=utf-8
Vary: Accept-Encoding
X-Content-Type-Options: nosniff
Date: Tue, 17 Jun 2025 00:22:41 GMT
Content-Length: 19

404 page not found
```

## Checklist
- [ ] Run `make build`
- [ ] Run `make generate`
- [ ] Update the current version in the `/VERSION` file. First line should be JUST the version e.g. `v10.13.1`. Then a blank line. Then release notes.


[SHAPI-1697]: https://clever.atlassian.net/browse/SHAPI-1697?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

maddy-jo committed:
Expose a Register function that can be used to install routes in subrouter

Takes the existing logic for generating the routes in gen-go/server and applies
them to an injected router. This allows a server to compose routers by calling

```
Register(parentRouter.PathPrefix("/subrouter-prefix").Subrouter(), controller)
```

Testing this hypothesis out in app-district-service.

maddy-jo committed:
Add -subrouter argument and cut out unnecessary server code when set

maddy-jo committed:
Enable configuring subrouters in parent swagger.yml with x-routers

The extension schema is of the form

```yaml
x-routers:
  - key: subrouter1
    path: /v1/abc
  - key: subrouter2
    path: /v1/xyz
```

This pattern assumes that the repo has a routers/ directory with subdirectories
routers/subrouter1 and routers/subrouter2, with both being subdirectories having
wag-generated route handlers.

maddy-jo committed:
Export NoOpLogger from logging/wagclientlogger package for tests

maddy-jo committed:
Update router generation not to include `basePath` value if generating subrouter

This change allows us to use the existing notion of `basePath` to generate the
correct client paths for a subrouter, assuming that we do not want to support
path parameters for a subrouter, which keeps us spec compliant for OAS 2.

maddy-jo committed:
Validate that parent x-routers[KEY].path matches subrouter basePath

maddy-jo committed:
Generate root client that uses implementation of subrouter clients

maddy-jo committed:
Document subrouters experiment in README and docs/subrouters.md

andruwm commented on logging/wagclientlogger/nooplogger.go:
```diff
@@ -0,0 +1,7 @@
+package wagclientlogger
+
+type NoOpLogger struct{}
```
We should have had one of these a long time ago.

maddy-jo replied:
yeah I think this would be worth cherry-picking into its own PR, probably?

andruwm replied:
Agreed

andruwm reviewed (commented):
The code changes here look reasonable. It's be helpful to add another swagger file which uses this new feature in `samples` so that its easier to understand how both the client and server look on a simple app.

andruwm commented:
> The code changes here look reasonable. It's be helpful to add another swagger file which uses this new feature in `samples` so that its easier to understand how both the client and server look on a simple app.

I just looked through the sample PR provided and was able to get a good idea of it there, but having it all in one place would be awesome at some point. Doesnt need to be now.

maddy-jo committed:
Generate JavaScript client methods for subrouter operations

maddy-jo committed:
fix: Generate TypeScript models from subrouter specs in addition to root spec

maddy-jo committed:
chore: Don't generate JavaScript clients for subrouters

Since the strategy for subrouter JavaScript clients (at least, so far), has been
to include the subrouters in the root client generation — which is the opposite
of how the Go clients are generated — we don't need to generate the subrouter
clients. This change autosets conf.generateJSClient to false if conf.subrouter
is true.

maddy-jo committed:
fix: Generate JS error classes from subrouter specs
